lib/IOExt/IOExt.cpp

```diff
@@ -2,101 +2,242 @@
+        CarStatePin pin = carState.pins[idx];
+        carState.idxOfPin.insert(pair<string, int>{pin.name, idx});
+        xSemaphoreTake(i2cBus.mutex, portMAX_DELAY);
+        ioExt.IOExtDevs[devNr].pinMode(pinNr, pin.mode);
+        xSemaphoreGive(i2cBus.mutex);
+        pin.value = pin.oldValue = 1;
+        sprintf(msg, "  0x%02x [%02d], mode:%s, value=%d (%s)\n", pin.port, carState.getIdx(pin.name),
+                pin.mode != OUTPUT ? "INPUT " : "OUTPUT", pin.value, pin.name.c_str());
+        printf(msg);
+        // dd.print(msg);
+      }
+      sprintf(msg, "[v] %s[%d] initialized.\n", getName().c_str(), devNr);
+      printf(msg);
+      dd.print(msg);
     }
   }
-
-  xSemaphoreGive(i2cBus.mutex);
-  // CRITICAL SECTION I2C: end
 }
 
 void IOExt::exit(void) {
   // TODO
 }
 
+CarStatePin CarState::pins[] = { // IOExtDev0
+    {0x00, INPUT_PULLUP, 1, 1, 0l, PinBatOnOff, batteryOnOffHandler},
+    {0x01, INPUT_PULLUP, 1, 1, 0l, PinPvOnOff, pvOnOffHandler},
+    {0x02, INPUT_PULLUP, 1, 1, 0l, PinMcOnOff, mcOnOffHandler},
+    {0x03, INPUT_PULLUP, 1, 1, 0l, PinEcoPower, ecoPowerHandler},
+    {0x04, INPUT_PULLUP, 1, 1, 0l, PinFwdBwd, fwdBwdHandler},
+    {0x05, INPUT_PULLUP, 1, 1, 0l, PinDUMMY06, NULL},
+    {0x06, INPUT_PULLUP, 1, 1, 0l, PinDUMMY07, NULL},
+    {0x07, OUTPUT, 1, 1, 0l, PinRelais11, NULL},
+    // IOExtDev1
+    {0x10, OUTPUT, 1, 1, 0l, PinRelais21, NULL},
+    {0x11, OUTPUT, 1, 1, 0l, PinRelais22, NULL},
+    {0x12, OUTPUT, 1, 1, 0l, PinRalais12, NULL},
+    {0x13, OUTPUT, 1, 1, 0l, PinRelais31, NULL},
+    {0x14, OUTPUT, 1, 1, 0l, PinRelais32, NULL},
+    {0x15, INPUT_PULLUP, 1, 1, 0l, PinBreakPedal, breakPedalHandler},
+    {0x16, INPUT_PULLUP, 1, 1, 0l, PinDUMMY19, NULL},
+    {0x17, INPUT_PULLUP, 1, 1, 0l, PinDUMMY17, NULL},
+    // IOExtDev2
+    {0x20, INPUT_PULLUP, 1, 1, 0l, PinIndicatorLeft, indicatorHandler},
+    {0x21, INPUT_PULLUP, 1, 1, 0l, PinIndicatorRight, indicatorHandler},
+    {0x22, INPUT_PULLUP, 1, 1, 0l, PinDriveLight, driveLightHandler},
+    {0x23, INPUT_PULLUP, 1, 1, 0l, PinLight, lightHandler},
+    {0x24, INPUT_PULLUP, 1, 1, 0l, PinHorn, hornHandler},
+    {0x25, INPUT_PULLUP, 1, 1, 0l, PinNextScreen, nextScreenHandler},
+    {0x26, INPUT_PULLUP, 1, 1, 0l, PinConstantMode, constantModeHandler},
+    {0x27, INPUT_PULLUP, 1, 1, 0l, PinConstantSet, constantSetHandler},
+    // IOExtDev3
+    {0x30, INPUT_PULLUP, 1, 1, 0l, PinDUMMY31, NULL},
+    {0x31, INPUT_PULLUP, 1, 1, 0l, PinReserve1, NULL},
+    {0x32, INPUT_PULLUP, 1, 1, 0l, PinDUMMY33, NULL},
+    {0x33, INPUT_PULLUP, 1, 1, 0l, PinDUMMY34, NULL},
+    {0x34, INPUT_PULLUP, 1, 1, 0l, PinDUMMY35, NULL},
+    {0x35, INPUT_PULLUP, 1, 1, 0l, PinDUMMY36, NULL},
+    {0x36, INPUT_PULLUP, 1, 1, 0l, PinDUMMY37, NULL},
+    {0x37, INPUT_PULLUP, 1, 1, 0l, PinDUMMY38, NULL}};
+
 void IOExt::handleIoInterrupt() {
-  // CRITICAL SECTION I2C: start
   xSemaphoreTake(i2cBus.mutex, portMAX_DELAY);
 
+#ifdef USE_LEGACY_PIN_READ
```

Comment:
I think that you deleted this macro? Does the project still compile?

Reply:
it compiles :(
the code is removed in the branch feature add-engineer-display